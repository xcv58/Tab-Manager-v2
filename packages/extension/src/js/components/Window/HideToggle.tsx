import React from 'react'
import IconButton from '@mui/material/IconButton'
import KeyboardArrowDown from '@mui/icons-material/KeyboardArrowDown'
import KeyboardArrowRight from '@mui/icons-material/KeyboardArrowRight'

export default ({ hide, toggleHide }) => {
  const icon = hide ? <KeyboardArrowRight /> : <KeyboardArrowDown />
  return (
    <IconButton
      onClick={toggleHide}
      className="focus:outline-none"
      aria-label="Toggle window hide"
    >
      {icon}
    </IconButton>
  )
}
