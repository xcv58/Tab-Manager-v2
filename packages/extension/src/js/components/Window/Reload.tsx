import React from 'react'
import Tooltip from '@material-ui/core/Tooltip'
import Refresh from '@material-ui/icons/Refresh'
import IconButton from '@material-ui/core/IconButton'

export default ({ reload }) => (
  <Tooltip title='Reload all tabs'>
    <IconButton onClick={reload} className='focus:outline-none'>
      <Refresh />
    </IconButton>
  </Tooltip>
)
