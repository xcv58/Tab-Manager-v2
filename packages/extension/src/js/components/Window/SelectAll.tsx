import React from 'react'
import { observer } from 'mobx-react-lite'
import Checkbox from '@mui/material/Checkbox'
import Tooltip from '@mui/material/Tooltip'

export default observer(({ win }) => {
  const { allTabSelected, someTabSelected, disableSelectAll, toggleSelectAll } =
    win
  const title = `${allTabSelected ? 'Unselect' : 'Select'} all tabs`
  return (
    <Tooltip title={title}>
      <Checkbox
        color="primary"
        inputProps={{
          'aria-label': title,
        }}
        disabled={disableSelectAll}
        checked={allTabSelected}
        onChange={(e) => {
          e.target.blur()
          toggleSelectAll()
        }}
        indeterminate={someTabSelected || disableSelectAll}
      />
    </Tooltip>
  )
})
