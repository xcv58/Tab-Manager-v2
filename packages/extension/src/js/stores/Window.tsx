import { action, computed, observable, makeObservable } from 'mobx'
import Tab from './Tab'
import { browser } from 'libs'
import Store from 'stores'
import log from 'libs/log'
import Focusable from './Focusable'

export default class Window extends Focusable {
  store: Store

  constructor(win, store: Store) {
    super(store)

    makeObservable(this, {
      tabs: observable,
      showTabs: observable,
      type: observable,
      title: observable,
      getTitle: action,
      setTitle: action,
      activate: action,
      hide: computed,
      visibleLength: computed,
      lastFocused: computed,
      canDrop: computed,
      invisibleTabs: computed,
      disableSelectAll: computed,
      matchedTabs: computed,
      allTabSelected: computed,
      someTabSelected: computed,
      add: action,
      remove: action,
      removeTabs: action,
      reload: action,
      close: action,
      toggleSelectAll: action,
      onMoved: action,
      onDetached: action,
      onAttched: action,
      toggleHide: action,
    })

    this.store = store
    Object.assign(this, win)
    this.tabs = win.tabs.map((tab) => new Tab(tab, store, this))
    // TODO: Remove this when we add concurrent mode
    this.showTabs = !this.store.windowStore.initialLoading
    this.getTitle()
  }

  tabs: Tab[] = []

  // TODO: Remove this when we add concurrent mode
  showTabs = false

  type = ''

  title = ''

  activate = () => {
    browser.windows.update(this.id, { focused: true })
  }

  get key() {
    return `win-${this.id}`
  }

  get hide() {
    return this.store.hiddenWindowStore.hiddenWindows[this.id]
  }

  get visibleLength() {
    if (this.hide) {
      return 2
    }
    const { length } = this.tabs.filter((x) => x.isVisible)
    return length ? length + 2 : length
  }

  get lastFocused() {
    return this.id === this.store.windowStore.lastFocusedWindowId
  }

  get canDrop() {
    return !['popup', 'devtools'].includes(this.type)
  }

  get invisibleTabs() {
    if (this.hide) {
      return this.tabs
    }
    return this.tabs.filter((x) => !x.isVisible)
  }

  get disableSelectAll() {
    if (this.hide) {
      return true
    }
    return this.matchedTabs.length === 0
  }

  get matchedTabs() {
    if (this.hide) {
      return []
    }
    return this.tabs.filter((x) => x.isMatched)
  }

  get allTabSelected() {
    return (
      !this.disableSelectAll &&
      this.matchedTabs.every(this.store.tabStore.isTabSelected)
    )
  }

  get someTabSelected() {
    return (
      !this.allTabSelected && this.tabs.some(this.store.tabStore.isTabSelected)
    )
  }

  getTab = (index: number) => {
    if (index < 0 || index >= this.tabs.length) {
      return null
    }
    return this.tabs[index]
  }

  add = (tab: Tab, index: number) => {
    if (index < 0 || index > this.tabs.length + 1) {
      throw new Error(`[Window-Store.add] get invalid index: "${index}"!`)
    }
    this.tabs.splice(index, 0, tab)
  }

  remove = (tab: Tab) => {
    const index = this.tabs.findIndex((x) => x.id === tab.id)
    if (index !== -1) {
      this.tabs.splice(index, 1)
    } else {
      throw new Error(
        `[Window-Store.remove] get invalid tab: ${JSON.stringify(tab)}!`,
      )
    }
  }

  removeTabs = (set) => {
    for (let index = 0; index < this.tabs.length; ) {
      const id = this.tabs[index].id
      if (set.has(id)) {
        this.tabs.splice(index, 1)
        set.delete(id)
      } else {
        index++
      }
    }
  }

  reload = () => {
    this.tabs.forEach((tab) => tab.reload())
  }

  close = () => {
    browser.windows.remove(this.id)
  }

  closeWindow = this.close

  toggleSelectAll = () => {
    log.debug('toggleSelectAll')
    const { allTabSelected, matchedTabs } = this
    const { selectAll, unselectAll } = this.store.tabStore
    if (allTabSelected) {
      unselectAll(matchedTabs)
    } else {
      selectAll(matchedTabs)
    }
  }

  select = this.toggleSelectAll

  onMoved = (tabId: number, moveInfo) => {
    const { fromIndex, toIndex } = moveInfo
    const toTab = this.getTab(toIndex)
    if (!toTab) {
      return false
    }
    if (toTab.id === tabId) {
      return true
    }
    const fromTab = this.getTab(fromIndex)
    if (!fromTab || fromTab.id !== tabId) {
      return false
    }
    this.tabs[fromIndex] = toTab
    this.tabs[toIndex] = fromTab
    return true
  }

  onDetached = (tabId: number, detachInfo) => {
    const { oldPosition } = detachInfo
    const oldTab = this.getTab(oldPosition)
    if (oldTab && oldTab.id === tabId) {
      this.tabs.splice(oldPosition, 1)
    }
  }

  onAttched = async (tabId: number, attachInfo) => {
    const { newPosition } = attachInfo
    const tab = this.getTab(newPosition)
    if (tab && tab.id === tabId) {
      return
    }
    const tabInfo = await browser.tabs.get(tabId)
    if (!tabInfo) {
      return false
    }
    this.tabs.splice(newPosition, 0, new Tab(tabInfo, this.store, this))
  }

  toggleHide = () => {
    if (this.hide) {
      this.store.hiddenWindowStore.showWindow(this.id)
    } else {
      this.store.hiddenWindowStore.hideWindow(this.id)
      this.store.focusStore.focus(this)
    }
  }

  getTitle = async () => {
    const res = await browser.storage.local.get({ [this.key]: '' })
    const title = res[this.key]
    console.log('key:', res, title)
    this.title = title
  }

  setTitle = async (title: string) => {
    this.title = title
    await browser.storage.local.set({ [this.key]: title })
  }
}
