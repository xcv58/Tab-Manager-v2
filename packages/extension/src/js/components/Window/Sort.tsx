import React from 'react'
import { observer } from 'mobx-react-lite'
import Tooltip from '@mui/material/Tooltip'
import SortIcon from '@mui/icons-material/Sort'
import IconButton from '@mui/material/IconButton'
import { useStore } from 'components/hooks/useStore'
import { WinProps } from 'components/types'

export default observer((props: WinProps) => {
  const { arrangeStore } = useStore()
  const { id } = props.win
  const { sortTabs } = arrangeStore
  return (
    <Tooltip title="Sort tabs">
      <IconButton onClick={() => sortTabs(id)} className="focus:outline-none">
        <SortIcon />
      </IconButton>
    </Tooltip>
  )
})
