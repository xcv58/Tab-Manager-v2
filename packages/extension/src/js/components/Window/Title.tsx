import React, { useRef, useEffect, useState } from 'react'
import { observer } from 'mobx-react-lite'
import SelectAll from 'components/Window/SelectAll'
import Sort from 'components/Window/Sort'
import CloseButton from 'components/CloseButton'
import { getNoun } from 'libs'
import classNames from 'classnames'
import Reload from './Reload'
import HideToggle from './HideToggle'
import { WinProps } from 'components/types'
import useReduceMotion from 'libs/useReduceMotion'

export const Title = (props) => {
  const { title, text, invisibleIndicator, hide } = props

  return (
    <h5 className="flex-auto text-2xl whitespace-nowrap">
      {title ? title : text} {!hide && invisibleIndicator}
    </h5>
  )
}
export const EditableTitle = observer((props: WinProps) => {
  const { win } = props
  const { tabs, invisibleTabs, hide, title, setTitle } = win
  const { length } = tabs
  const text = `${length} ${getNoun('tab', length)}`
  const invisibleLength = invisibleTabs.length
  const invisibleIndicator =
    invisibleLength > 0 && `/ ${invisibleLength} hidden`

  const [editing, setEditing] = useState<boolean>(false)
  const [value, setValue] = useState<string>(title || text)
  useEffect(() => {
    setValue(title || text)
  }, [title, text])

  return (
    <div
      className="flex-auto overflow-hidden text-base text-left rounded-sm"
      onClick={() => {
        setEditing((x) => !x)
      }}
    >
      {editing ? (
        <input
          autoFocus
          onChange={(e) => {
            setValue(e.target.value)
          }}
          onKeyDown={(e) => {
            if (e.key === 'Enter') {
              e.preventDefault()
              setTitle(value)
              setEditing(false)
            } else if (e.key === 'Escape') {
              e.preventDefault()
              setEditing(false)
            }
          }}
          value={value}
          onFocus={(e) => e.currentTarget.select()}
        />
      ) : (
        <Title {...{ text, title, invisibleIndicator, hide }} />
      )}
    </div>
  )
})

export default observer((props: WinProps & { className: string }) => {
  const nodeRef = useRef(null)
  const { className, win } = props
  const { reload, hide, toggleHide, isFocused } = win
  const reduceMotion = useReduceMotion()
  useEffect(() => {
    if (isFocused) {
      nodeRef.current.focus({ preventScroll: true })
      nodeRef.current.scrollIntoView({
        behavior: reduceMotion ? 'auto' : 'smooth',
      })
    }
  }, [isFocused])
  useEffect(() => {
    win.setNodeRef(nodeRef)
  })
  return (
    <div
      tabIndex={-1}
      ref={nodeRef}
      className={classNames(
        'flex justify-between items-center font-bold border-0',
        className,
      )}
    >
      <SelectAll {...props} />
      <EditableTitle {...props} />
      {!hide && (
        <>
          <Sort {...props} />
          <Reload {...{ reload }} />
        </>
      )}
      <CloseButton onClick={() => props.win.close()} />
      <HideToggle {...{ hide, toggleHide }} />
    </div>
  )
})
