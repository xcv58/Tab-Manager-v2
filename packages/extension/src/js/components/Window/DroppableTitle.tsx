import React from 'react'
import { observer } from 'mobx-react-lite'
import { useDrop } from 'react-dnd'
import DropIndicator from 'components/DropIndicator'
import { ItemTypes, getTargetTab } from 'libs/react-dnd'
import classNames from 'classnames'
import Title from './Title'
import { useStore } from 'components/hooks/useStore'
import { WinProps } from 'components/types'

export default observer((props: WinProps) => {
  const { win } = props
  const { dragStore } = useStore()
  const [dropProps, drop] = useDrop({
    accept: ItemTypes.TAB,
    canDrop: () => win.canDrop,
    drop: (_, monitor) => {
      if (monitor.didDrop()) {
        return
      }
      const tab = getTargetTab(win.tabs, true)
      if (tab) {
        dragStore.drop(tab, true)
      }
    },
    collect: (monitor) => {
      return {
        canDrop: monitor.canDrop(),
        isDragging: !!monitor.getItem(),
        isOver: monitor.isOver({ shallow: true }),
      }
    },
  })
  const { isOver, canDrop, isDragging } = dropProps
  const preview = canDrop && isOver && <DropIndicator />
  return (
    <div ref={drop}>
      <Title
        {...props}
        className={classNames({
          'bg-red-500': isDragging && isOver && !canDrop,
        })}
      />
      {preview}
    </div>
  )
})
