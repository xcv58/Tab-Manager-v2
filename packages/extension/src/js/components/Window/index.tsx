import React from 'react'
import classNames from 'classnames'
import { observer } from 'mobx-react-lite'
import { useDrop } from 'react-dnd'
import DroppableTitle from './DroppableTitle'
import Tabs from './Tabs'
import DropIndicator from 'components/DropIndicator'
import { ItemTypes, getTargetTab } from 'libs/react-dnd'
import { useStore } from 'components/hooks/useStore'
import { CSSProperties } from '@material-ui/styles'
import Loading from 'components/Loading'
import { TAB_HEIGHT } from 'libs'
import { WinProps } from 'components/types'

export default observer((props: WinProps & { width: string }) => {
  const { dragStore, userStore } = useStore()
  const { win, width } = props
  const { lastFocused, showTabs, visibleLength } = win
  const [dropProps, drop] = useDrop({
    accept: ItemTypes.TAB,
    canDrop: () => win.canDrop,
    drop: (_, monitor) => {
      if (monitor.didDrop()) {
        return
      }
      const tab = getTargetTab(win.tabs, false)
      if (tab) {
        dragStore.drop(tab, false)
      }
    },
    collect: (monitor) => {
      return {
        canDrop: monitor.canDrop(),
        isDragging: !!monitor.getItem(),
        isOver: monitor.isOver({ shallow: true })
      }
    }
  })
  const { canDrop, isOver, isDragging } = dropProps
  const style: CSSProperties = {
    minWidth: `${userStore.tabWidth}rem`,
    width,
    height: 'fit-content',
    boxSizing: 'border-box'
  }
  const dropIndicator = canDrop && isOver && <DropIndicator />
  if (!win.visibleLength) {
    return null
  }
  return (
    <div
      ref={drop}
      style={style}
      className={classNames('window', {
        'bg-red-500': isDragging && isOver && !canDrop
      })}
    >
      <div
        className={classNames({
          'shadow-2xl': lastFocused,
          'shadow-sm hover:shadow-lg': !lastFocused
        })}
      >
        <DroppableTitle {...props} />
        {showTabs && <Tabs {...props} />}
        {!showTabs && (
          <div style={{ height: TAB_HEIGHT * (visibleLength - 2) }}>
            <Loading small />
          </div>
        )}
        {dropIndicator}
      </div>
    </div>
  )
})
