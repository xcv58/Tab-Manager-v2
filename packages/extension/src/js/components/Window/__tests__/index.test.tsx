import React from 'react'
import { connectDropTarget } from 'test'
import { DndProvider } from 'react-dnd'
import { HTML5Backend } from 'react-dnd-html5-backend'
import Window from 'components/Window'
import { createTheme } from '@mui/material/styles'
import { render } from '@testing-library/react'

jest.mock('../Tabs', () => 'tabs')

const theme = createTheme()
const tabs = [{ id: 1 }, { id: 2 }]
const props = {
  connectDropTarget,
  dragStore: {
    drop: jest.fn(),
  },
  win: {
    tabs,
    showTabs: true,
  },
  width: '100%',
  theme,
}

describe('Window', () => {
  it('render correct components', () => {
    const { container } = render(
      <DndProvider backend={HTML5Backend}>
        <Window {...props} />
      </DndProvider>
    )
    expect(container).toMatchSnapshot()
  })

  it.skip('render error.light backgroundColor if canDrop is false', () => {
    const { container } = render(
      <DndProvider backend={HTML5Backend}>
        <Window {...props} isDragging isOver canDrop={false} />
      </DndProvider>
    )
    expect(container).toMatchSnapshot()
    // expect(el.find(Title).length).toBe(1)
    // expect(el.find(Tabs).length).toBe(1)
    // expect(el.find('div').props().style.backgroundColor).toBe(
    //   theme.palette.error.light
    // )
  })

  it.skip('render Preview based on canDrop & isOver', () => {
    const { container } = render(<Window {...props} canDrop isOver />)
    expect(container).toMatchSnapshot()

    // expect(el.find(Preview).length).toBe(0)
    // el = render(<Window {...props} isOver />)
    // expect(el.find(Preview).length).toBe(0)
    // el = render(<Window {...props} canDrop />)
    // expect(el.find(Preview).length).toBe(0)
    // el = render(<Window {...props} canDrop isOver />)
    // expect(el.find(Preview).length).toBe(1)
  })

  // it.skip('render correct elevation based on lastFocused', () => {
  //   let el = render(<Window {...props} />)
  //   expect(el.find(Paper).props().elevation).toBe(2)
  //   el = render(<Window {...props} win={{ lastFocused: true }} />)
  //   expect(el.find(Paper).props().elevation).toBe(16)
  // })
})
