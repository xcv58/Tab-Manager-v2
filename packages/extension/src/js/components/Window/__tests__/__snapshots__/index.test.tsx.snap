// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Window render correct components 1`] = `<div />`;

exports[`Window render error.light backgroundColor if canDrop is false 1`] = `<div />`;
